src/xcode/ENA/ENA/Source/Services/RecycleBin/RecycleBin.swift

```diff
@@ -0,0 +1,207 @@
+	func hash(into hasher: inout Hasher) {
+		hasher.combine(item)
+	}
+}
+
+enum RestorationError: Error {
+	case testError(TestRestorationError)
+	case certificateError(CertificateRestorationError)
+}
+
+enum TestRestorationError: Error {
+	case some
+}
+
+protocol TestRestorationHandling {
+	var canRestore: ((CoronaTest) -> Result<Void, TestRestorationError>) { get set }
+	var restore: ((CoronaTest) -> Void) { get set }
+}
+
+enum CertificateRestorationError: Error {
+	case some
+}
+
+protocol CertificateRestorationHandling {
+	var canRestore: ((HealthCertificate) -> Result<Void, CertificateRestorationError>) { get set }
+	var restore: ((HealthCertificate) -> Void) { get set }
+}
+
+class RecycleBin {
+
+	// MARK: - Init
+
+	init(
+		store: RecycleBinStoring
+	) {
+		self.store = store
+	}
+
+	// MARK: - Internal
+
+	var testRestorationHandler: TestRestorationHandling!
+	var certificateRestorationHandler: CertificateRestorationHandling!
+
+	/// This function moves an item into the bin.
+	///
+	/// - Parameter item: The item which needs to be moved into the bin.
+	/// - Parameter recycledAt: Date at which the item was moved into the bin. Default value: `Date()`
+	/// - Returns: A `RecycleBinItem` which contains the item moved into the bin and metadata added from the bin (like `recycledAt`).
+	@discardableResult
+	func moveToBin(_ item: RecycledItem, recycledAt: Date = Date()) -> RecycleBinItem {
+		Log.info("Recycle item.", log: .recycleBin)
+
+		let binItem = RecycleBinItem(
+			recycledAt: recycledAt,
+			item: item
+		)
+		store.recycleBinItems.insert(binItem)
+		return binItem
+	}
+
+	/// Asks if an item can be restored. This call is not restoring the item immediately.
+	/// Instead its asking the restoration handler, if the item can be restored.
+	///	In some cases the item cannot be restored immediately and needs some further processing or a user decision.
+	///	Please see `restore(...)` for restoring the item immediately (Call it after further processing was done and the item can finally be restored).
+	///
+	/// - Parameter item: The item which needs to be restored.
+	/// - Returns: `Result.Success`: The item can be restored. `Result.Failure`: Item cannot be restored and needs some further processing.
+	func canRestore(_ item: RecycleBinItem) -> Result<Void, RestorationError> {
+		Log.info("Ask for item restoration.", log: .recycleBin)
+
+		switch item.item {
+		case .certificate(let certificate):
+			Log.info("Ask for certificate item restoration.", log: .recycleBin)
+			let canRestoreResult = certificateRestorationHandler.canRestore(certificate)
+			return canRestoreResult.mapError { RestorationError.certificateError($0) }
+		case .coronaTest(let coronaTest):
+			Log.info("Ask for test item restoration.", log: .recycleBin)
+			let canRestoreResult = testRestorationHandler.canRestore(coronaTest)
```

Comment:
Is there a change to make the `testRestorationHandler ` optional and then call `testRestorationHandler?.canRestore(coronaTest)`?